File: go-concurrent/src/singleton.go
package src

import "sync"

type Singleton struct {
	money int
}

var instance *Singleton
var mut sync.Mutex

func (singleton *Singleton) GetMoney() int {
	singleton.money++
	return singleton.money
}

// GetInstance 使用双重检查锁实现单例模式
func GetInstance() *Singleton {
	if instance == nil {
		mut.Lock()
		defer mut.Unlock()
		if instance == nil {
			instance = &Singleton{
				money: 10_000,
			}
		}
	}
	return instance
}

File: design-pattern/src/creational/singleton.ts
// 创建型模式
// 单例模式: 保证一个类只有一个实例 (该实例全局可访问)
class Singleton {
  static #instance: Singleton;

  private constructor() {}

  public static get instance(): Singleton {
    if (!Singleton.#instance) {
      Singleton.#instance = new Singleton();
    }
    return Singleton.#instance;
  }
}

function clientCode() {
  const s1 = Singleton.instance;
  const s2 = Singleton.instance;
  console.log("s1 === s2:", s1 === s2); // true
}

clientCode();

export default {};
